Done

meg-gutshall opened the pull request:
Initial commit -- FINALLYlearn help

Commit:
Done.

meg-gutshall committed:
Done.

meg-gutshall committed:
Done.

meg-gutshall committed:
Done.

meg-gutshall committed:
Done.

meg-gutshall committed:
Initial commit -- FINALLYlearn help

meg-gutshall committed:
Refactored board.rb

meg-gutshall committed:
Refactored game.rb

meg-gutshall committed:
Started developing CLI

meg-gutshall committed:
Moved game prompt to game.rb and edited first prompt.

meg-gutshall committed:
Added more interaction to first CLI prompt.

meg-gutshall committed:
Create new players according to user input. Still need to figure out order and name.

meg-gutshall committed:
Broke out CLI methods in Game class

meg-gutshall committed:
Added @name attribute to Player class and edited tests to pass.

meg-gutshall committed:
Created GameCLI class.

meg-gutshall committed:
Defined @@num_players GameCLI class variable.

meg-gutshall committed:
Working on starting game play.

meg-gutshall committed:
Deleted GameCLI class -- everything back in Game class.

meg-gutshall committed:
Able to initialize a new game with the chosen players.

meg-gutshall committed:
Have to initialize new game in

meg-gutshall committed:
Format

meg-gutshall committed:
Fixed formatting for computer play.

meg-gutshall committed:
Working gamebin/tictactoe Just needs a few logic fixes and refactoring.

meg-gutshall committed:
Program exits after declining a replay.

meg-gutshall committed:
Created @@game class variable to hold game info

meg-gutshall committed:
Made winner display message include player name instead of player token.

meg-gutshall committed:
Start coding strategic moves

meg-gutshall committed:
First and second moves if the computer plays first.

meg-gutshall committed:
Need to revisit turn order methods

meg-gutshall committed:
First few moves done

meg-gutshall committed:
Refactored some methods

meg-gutshall committed:
Changed computer strategy

meg-gutshall committed:
Done.

meg-gutshall committed:
Redoing AI strategy

meg-gutshall committed:
Redoing AI strategy

meg-gutshall committed:
Added logic for token X second move

meg-gutshall committed:
Added logic for token X third move

meg-gutshall committed:
Added logic for token O first move

meg-gutshall committed:
Added logic for token O second move

meg-gutshall committed:
Added logic by move

meg-gutshall committed:
Done.

meg-gutshall committed:
Finished

meg-gutshall committed:
Added turn methods into

meg-gutshall committed:
Added game argument to

meg-gutshall committed:
Need to find out how to return correct move value

meg-gutshall committed:
Still working on AI logic

meg-gutshall committed:
No AI logic BUT IT WORKSbin/tictactoe

meg-gutshall committed:
Done.

meg-gutshall committed:
Attempted AI logic commented out

meg-gutshall committed:
Attempted AI logic commented out

kkterai commented on lib/cli.rb:
```diff
@@ -0,0 +1,83 @@
+class CLI
```
Good job encapsulating your CLI logic into a class!

kkterai commented on lib/board.rb:
```diff
@@ -0,0 +1,45 @@
+  attr_accessor :cells
+
+  def initialize
+    @cells = Array.new(9, " ")
```
I would just call `self.reset!` here and delegate to that method to create the default empty board.

kkterai commented on lib/board.rb:
```diff
@@ -0,0 +1,45 @@
+  end
+
+  def reset!
+    self.cells = Array.new(9, " ")
```
Nice job using Array.new(9," ") - this will result in fewer errors.

kkterai commented on lib/board.rb:
```diff
@@ -0,0 +1,45 @@
+  end
+
+  def full?
+    !cells.include?(" ")
```
Use the most expressive and efficient iterator for a problem. `Board#full?` is asking if all elements are not empty strings. You can use logic that is symmetrical to those semantics with the [`#all?`](http://ruby-doc.org/core-2.2.3/Enumerable.html#method-i-all-3F) iterator.

`cells.all? {|cell| cell != " "}`

This is also more efficient in that the loop will stop after the iteration for which a cell is not empty.

kkterai commented on lib/board.rb:
```diff
@@ -0,0 +1,45 @@
+  end
+
+  def turn_count
+    9 - cells.count(" ").to_i
```
Use the highest level and most expressive iterator possible to solve a problem. [`count`](http://ruby-doc.org/core-2.2.0/Array.html#method-i-count) on the cells is perfect and the rest of this method is not needed. Entire body could read:

```ruby
def turn_count
  cells.count { |cell| cell == "X" || cell == "O" }
end
```

meg-gutshall replied:
Didn't need #to_i but definitely need to keep the rest of the line.

kkterai commented on lib/board.rb:
```diff
@@ -0,0 +1,45 @@
+  end
+
+  def valid_move?(input)
+    input.to_i.between?(1, 9) && !taken?(input)
```
Nice job on this method! Not only did you use the `#between?` method, I like how you used the 'and' operator (`&&`) to combine multiple logical expressions and utilize their implicit return value.

kkterai commented on lib/game.rb:
```diff
@@ -0,0 +1,85 @@
+    self
+  end
+
+  def current_player
```
Nice job using [`Integer#even?`](http://ruby-doc.org/core-1.8.7/Integer.html#method-i-even-3F) here. It's a very semantic, readable method. :-)

kkterai commented on lib/game.rb:
```diff
@@ -0,0 +1,85 @@
+    end
+    win
+  end
+
```
Nice job working out the logic to a difficult problem!  Good idea to abstract out the logic into two methods.

As another alternative, you can use [`Enumerable#detect`](http://ruby-doc.org/core-2.2.3/Enumerable.html#method-i-detect) to return the winning combination or `nil` if none is found.

```ruby
  def won?
    WIN_COMBINATIONS.detect do |combo|
      @board.cells[combo[0]] == @board.cells[combo[1]] &&
      @board.cells[combo[0]] == @board.cells[combo[2]] &&
      @board.taken?(combo[0] + 1)
    end
  end
```

meg-gutshall replied:
Thanks! I had help from @DRBragg with abstracting the method.  I was pretty frustrated with this method though. I believe his words were "You look like you're about to punch a child in the face."

kkterai commented on lib/game.rb:
```diff
@@ -0,0 +1,85 @@
+
+  def winner
+    board.cells[won?&.first] if won?
+  end
```
Instead of calling `Game#won?` twice, consider using an if statement assignment. For example:

```ruby
if won = won?
  board.cells[won.first]
end
```

If `won?` returns true, then and only then will the local variable `won` be assigned, otherwise, the method will just abort with a false return value. But if `won` is assigned, we can use that variable instead of calling `won?` again within the if statement.

meg-gutshall committed:
Refactor Board.new method

meg-gutshall committed:
Refactor Human class #move

meg-gutshall committed:
Refactor Board#turn_count

meg-gutshall committed:
Refactor Board#full?

meg-gutshall committed:
Added other solution to Game#winner

meg-gutshall committed:
Update AI logic

meg-gutshall committed:
Adjust @player_move variable

Variable does not need to be an instance variable. It only needs to
return the optimal move.

meg-gutshall committed:
Add AI play strategy notes

meg-gutshall committed:
Add resource link

meg-gutshall committed:
Create and rename files for logic lab

meg-gutshall committed:
Run bundler

meg-gutshall committed:
Work on AI logic

meg-gutshall committed:
Pseudocoding AI logic

meg-gutshall committed:
Pseudocoding AI logic

meg-gutshall committed:
Fix markdown errors

meg-gutshall committed:
Test markdown style

meg-gutshall committed:
Fix Ruby code indentation

meg-gutshall committed:
Refactor won? method

meg-gutshall committed:
Work on AI logic

mguts-penn committed:
Lint code

mguts-penn committed:
Add to .gitignore file
